### DISCUSSION TITLE
Why does the tho syntaxes for the same not generate the same code?

### DISCUSSION THREAD

**Discussion by tndata** [dotnet/csharplang#2678] [Q&A]
I would expect the two methods below to generate the same code:

```
public class Product
{
    public int Id;
    public string Name;

    public void MethodA(out int id,
                        out string name) =>
     
      (id, name) = (Id, Name);

    public void MethodB(out int id,
                        out string name)
    {
        id = Id;
        name = Name;
    }
}
```

But in Sharplab the ValueTuple assignment in MethodA becomes

```
    public void MethodA(out int id, out string name)
    {
        int id2 = Id;
        string name2 = Name;
        id = id2;
        name = name2;
    }

    public void MethodB(out int id, out string name)
    {
        id = Id;
        name = Name;
    }
```

Bug or by design? The ValueTuple version seems a bit more inefficient?

**Comment by BhaaLseN** [comment]
I wouldn't be surprised if this is optimized out anyways.

But the reason is probably that this also allows you to swap values:
```c#
string left = "two";
string right = "one";
(left, right) = (right, left);
Console.WriteLine("{0} <-> {1}", left, right);
```

**Comment by canton7** [comment]
> I wouldn't be surprised if this is optimized out anyways.

[Nope](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxNMAfAAgZgARYBMeACggPYAmArgMYwCwAUAN7N4cH4CWAdjHgCSlANztOuAgEYADHgByEMHDFNxHSVgAseALJwYACyoBBABTlqAvtcpJ1nR0+edLArLLy8lcAJR4AXgA+BzxQvDNuOy8ffwCI4SQFWNUHTR19IyoAIQsrPBsCu3CXUtd8jzlvZV8HNiZSqMChURKY5WbFZVVHAF9mXqA==)

**Comment by mburbea** [comment]
The optimization seems to apply if you are setting a `readonly` field or a getter only property in a constructor.
See [sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABAzAAgwJhwBUBXABwBsYATAWACgBvBnVvXASwDsAXHAQRyMA5jB4BuAL4s22PAEYADDgBCQ0ROn027YuSrUAFNz4BDADQLlwAJQ4AvAD4ch/pZU37LizlviGWgxyBDgActAAtqYUDMzaspy8AupiUjKswUqqKZrpuuFQURTGSRYYWbZ5cTo6gl6m4jh5OmpewP7xrFqSbAx5QA===)

**Comment by ufcpp** [comment]
Evaluation order is not the same:

```cs
using System;

class X
{
    public int A { get { Console.WriteLine("X.A"); return 0; } }
    public int B { get { Console.WriteLine("X.B"); return 0; } }
}

class Y
{
#if tuple
    // Y.A → Y.B → X.A → X.B
    public Y(X other) => (A, B) = (other.A, other.B);
#else
    // Y.A → X.A → Y.B → X.B
    public Y(X other)
    {
        A = other.A;
        B = other.B;
    }
#endif

    private int _a;
    private int _b;
    public ref int A { get { Console.WriteLine("Y.A"); return ref _a; } }
    public ref int B { get { Console.WriteLine("Y.B"); return ref _b; } }
}
```

**Comment by canton7** [comment]
@ufcpp While that's true, it's not the same code that the OP posted, so I'm not sure that it's relevant?

**Comment by tndata** [comment]
Even the x86 assembly and IL code in Sharplab for the original code is different. So I was just curious why there's a difference in Sharplab.

**Comment by canton7** [comment]
> So I was just curious why there's a difference in Sharplab.

It seems that when fields are accessed as part of a tuple, those fields are always copied to locals. It seems there are some cases that nobody's bothered to optimize - your example is one, this is another:

```csharp
using System;
public class Product
{
    public int Id;
    public string Name;

    public void MethodA() 
    {
        var (a, b) = (Id, Name);
        Console.WriteLine($"{a} {b}");
    }
}
```

Compiles to two sets of locals:

```csharp
public void MethodA()
{
    int num = Id;
    string text = nName;
    int num2 = num;
    string arg = text;
    Console.WriteLine(string.Format("{0} {1}", num2, arg));
}
```

I think the relevant code is [around here](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs,111).
